.github/workflows/release.yml

```diff
@@ -28,55 +45,88 @@ jobs:
           poetry env use python${{ matrix.python-version }}
           poetry install --verbose --no-root
 
-      # NOTE: This step will override, for this release only, the version set in the project. That is, the version will
-      #       be updated in the checked out source on the runner and used when building the package and release
-      #       artifacts but it will not be committed back to source control (e.g., in `main` branch). This means version
-      #       bumps in source control are not strictly enforced and syncronized with release versions.
-      # TODO: Consider a more fully automated continuous deployment pipeline. Some tools and workflows that can help are
-      #       using conventional commits, commitizen, commitlint, semantic-pull-requests, and python-semantic-release.
-      - name: Set the release version
-        run: poetry version ${GITHUB_REF#refs/tags/v}
-
-      # NOTE: Run the tests for the current active Python version, as a sanity check. It should uncover the situation
-      #       where the tag version doesn't adhere to the post-normalized PEP440 form: https://peps.python.org/pep-0440/
+      - name: Set phylum_original_version value
+        run: echo "phylum_original_version=$(poetry version --short)" >> $GITHUB_ENV
+
+      - name: Set to next version for build
+        run: |
+          if [ ${{ github.event.inputs.prerelease }} = "true" ]; then
+            poetry version $(poetry run semantic-release print-version --next --prerelease)
+          else
+            poetry version $(poetry run semantic-release print-version --next)
+          fi
+
+      - name: Set phylum_release_version value
+        run: echo "phylum_release_version=$(poetry version --short)" >> $GITHUB_ENV
+
+      # NOTE: Run the tests for the current active Python version, as a sanity check.
       - name: Run tox via poetry
         run: poetry run tox
 
       - name: Build wheel and source distribution
         run: poetry build -vvv
 
-      - name: Create GitHub release
-        id: github_release
-        uses: softprops/action-gh-release@v0.1.14
-        with:
-          # This check is already filtered down to only 'refs/tags/' and shouldn't "overmatch" for a different ref type.
-          # Look for the post-normalized PEP440 pre-release suffixes: https://peps.python.org/pep-0440/#pre-releases
-          prerelease: ${{ contains(github.ref, 'a') || contains(github.ref, 'b') || contains(github.ref, 'rc') }}
-          generate_release_notes: true
-          token: ${{ secrets.GITHUB_TOKEN }}
-          fail_on_unmatched_files: true
-          files: |
-            dist/*.whl
-            dist/*.tar.gz
-
-      - name: Publish to PyPI
-        run: poetry publish --username __token__ --password ${{ secrets.PYPI_API_TOKEN }}
+      - name: Publish to TestPyPI
+        id: publish_test_pypi
+        run: poetry publish --repository testpypi --username __token__ --password ${{ secrets.TESTPYPI_API_TOKEN }}
```

Comment:
This really feels like a misuse of TestPyPI....

> TestPyPI – a separate instance of the Python Package Index that allows you to try distribution tools and processes without affecting the real index

Using it as staging before a release does not seem to fit that description.

Reply:
There is disagreement here. The [User guide](https://packaging.python.org/) linked to from TestPyPI itself provides a full guide on [Publishing package distribution releases using GitHub Actions CI/CD workflows](https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/), which includes publishing the distribution to PyPI and TestPyPI:

> Now, whenever you push a tagged commit to your Git repository remote on GitHub, this workflow will publish it to PyPI. And it’ll publish any push to TestPyPI which is useful for providing test builds to your alpha users as well as making sure that your release pipeline remains healthy!

Other tutorials in the same user guide also mention this:

> Keep in mind that this tutorial showed you how to upload your package to Test PyPI, which isn’t a permanent storage. The Test system occasionally deletes packages and accounts. It is best to use TestPyPI for testing and experiments like this tutorial.

Using TestPyPI for "...trying distribution tools and processes..." and "testing" is exactly the point of this step in the `Release` workflow. It is better to know that the package works as intended _before_ publishing it to the production PyPI registry, where consumers of the package have a higher expectation of working software. This split will be even more useful when there is a more holistic E2E or integration (as it were) level testing suite in place to confirm proper behavior.